File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/drops/impl/EcoDropQueue.java
package com.willfp.eco.internal.drops.impl;

import com.willfp.eco.core.drops.DropQueue;
import com.willfp.eco.util.TelekinesisUtils;
import lombok.AccessLevel;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.ExperienceOrb;
import org.bukkit.entity.Player;
import org.bukkit.event.player.PlayerExpChangeEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.Vector;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;

public class EcoDropQueue extends DropQueue {
    @Getter(AccessLevel.PROTECTED)
    private final List<ItemStack> items;
    
    @Getter(AccessLevel.PROTECTED)
    private int xp;
    
    @Getter(AccessLevel.PROTECTED)
    private final Player player;

    @Getter(AccessLevel.PROTECTED)
    private Location loc;
    
    @Getter(AccessLevel.PROTECTED)
    private boolean hasTelekinesis = false;

    public EcoDropQueue(@NotNull final Player player) {
        super(player);
        this.items = new ArrayList<>();
        this.xp = 0;
        this.player = player;
        this.loc = player.getLocation();
    }

    @Override
    public DropQueue addItem(@NotNull final ItemStack item) {
        this.items.add(item);
        return this;
    }

    @Override
    public DropQueue addItems(@NotNull final Collection<ItemStack> itemStacks) {
        this.items.addAll(itemStacks);
        return this;
    }

    @Override
    public DropQueue addXP(final int amount) {
        this.xp += amount;
        return this;
    }

    @Override
    public DropQueue setLocation(@NotNull final Location location) {
        this.loc = location;
        return this;
    }

    @Override
    public DropQueue forceTelekinesis() {
        this.hasTelekinesis = true;
        return this;
    }

    @Override
    public void push() {
        if (!hasTelekinesis) {
            hasTelekinesis = TelekinesisUtils.testPlayer(player);
        }

        World world = loc.getWorld();
        assert world != null;
        loc = loc.add(0.5, 0.5, 0.5);

        items.removeIf(itemStack -> itemStack.getType() == Material.AIR);
        if (items.isEmpty()) {
            return;
        }

        if (hasTelekinesis) {
            HashMap<Integer, ItemStack> leftover = player.getInventory().addItem(items.toArray(new ItemStack[0]));
            for (ItemStack drop : leftover.values()) {
                world.dropItem(loc, drop).setVelocity(new Vector());
            }
            if (xp > 0) {
                PlayerExpChangeEvent event = new PlayerExpChangeEvent(player, xp);
                Bukkit.getPluginManager().callEvent(event);
                ExperienceOrb orb = (ExperienceOrb) world.spawnEntity(player.getLocation().add(0, 0.2, 0), EntityType.EXPERIENCE_ORB);
                orb.setVelocity(new Vector(0, 0, 0));
                orb.setExperience(event.getAmount());
            }
        } else {
            for (ItemStack drop : items) {
                world.dropItem(loc, drop).setVelocity(new Vector());
            }
            if (xp > 0) {
                ExperienceOrb orb = (ExperienceOrb) world.spawnEntity(loc, EntityType.EXPERIENCE_ORB);
                orb.setExperience(xp);
            }
        }
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/drops/impl/EcoDropQueue.kt
package com.willfp.eco.internal.drops.impl

import com.willfp.eco.core.drops.InternalDropQueue
import com.willfp.eco.util.TelekinesisUtils
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.EntityType
import org.bukkit.entity.ExperienceOrb
import org.bukkit.entity.Player
import org.bukkit.event.player.PlayerExpChangeEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.util.Vector

open class EcoDropQueue(player: Player) : InternalDropQueue {
    val items: MutableList<ItemStack>
    var xp: Int
    val player: Player
    var loc: Location

    private var hasTelekinesis = false
    override fun addItem(item: ItemStack): InternalDropQueue {
        items.add(item)
        return this
    }

    override fun addItems(itemStacks: Collection<ItemStack>): InternalDropQueue {
        items.addAll(itemStacks)
        return this
    }

    override fun addXP(amount: Int): InternalDropQueue {
        xp += amount
        return this
    }

    override fun setLocation(location: Location): InternalDropQueue {
        loc = location
        return this
    }

    override fun forceTelekinesis(): InternalDropQueue {
        hasTelekinesis = true
        return this
    }

    override fun push() {
        if (!hasTelekinesis) {
            hasTelekinesis = TelekinesisUtils.testPlayer(player)
        }
        val world = loc.world!!
        loc = loc.add(0.5, 0.5, 0.5)
        items.removeIf { itemStack: ItemStack -> itemStack.type == Material.AIR }
        if (items.isEmpty()) {
            return
        }
        if (hasTelekinesis) {
            val leftover = player.inventory.addItem(*items.toTypedArray())
            for (drop in leftover.values) {
                world.dropItem(loc, drop!!).velocity = Vector()
            }
            if (xp > 0) {
                val event = PlayerExpChangeEvent(player, xp)
                Bukkit.getPluginManager().callEvent(event)
                val orb =
                    world.spawnEntity(player.location.add(0.0, 0.2, 0.0), EntityType.EXPERIENCE_ORB) as ExperienceOrb
                orb.velocity = Vector(0, 0, 0)
                orb.experience = event.amount
            }
        } else {
            for (drop in items) {
                world.dropItem(loc, drop).velocity = Vector()
            }
            if (xp > 0) {
                val orb = world.spawnEntity(loc, EntityType.EXPERIENCE_ORB) as ExperienceOrb
                orb.experience = xp
            }
        }
    }

    init {
        items = mutableListOf()
        xp = 0
        this.player = player
        loc = player.location
    }
}
